File: ios/DesioraAI/Models/ApiModels.swift
import Foundation

// MARK: - Authentication Models
struct LoginRequest: Codable {
    let email: String
    let password: String
}

struct RegisterRequest: Codable {
    let email: String
    let password: String
    let username: String?
    let full_name: String?
    
    enum CodingKeys: String, CodingKey {
        case email, password, username
        case full_name
    }
}

struct TokenResponse: Codable {
    let access_token: String
    let refresh_token: String
    let token_type: String
}

struct RefreshTokenRequest: Codable {
    let refresh_token: String
}

struct UserResponse: Codable {
    let id: Int
    let email: String
    let username: String?
    let full_name: String?
    let is_active: Bool
    let is_verified: Bool
    let role: String
    let oauth_provider: String
    let created_at: String
    let updated_at: String?
}

struct OAuthRequest: Codable {
    let code: String
    let provider: String
    let redirect_uri: String?
}

// MARK: - Image Models
struct PresignUploadRequest: Codable {
    let filename: String
    let content_type: String
    let file_size: Int
}

struct PresignUploadResponse: Codable {
    let upload_url: String
    let s3_key: String
    let image_id: Int
    let expires_in: Int
}

struct ConfirmUploadRequest: Codable {
    let s3_key: String
}

// MARK: - Design Job Models
struct DesignJobCreateRequest: Codable {
    let job_type: String
    let prompt: String
    let parameters: [String: String]?
}

struct DesignJobResponse: Codable, Identifiable {
    let id: Int
    let user_id: Int
    let job_type: String
    let prompt: String
    let status: String
    let parameters: [String: String]?
    let result_urls: [String]?
    let result_metadata: [String: String]?
    let error_message: String?
    let retry_count: Int
    let max_retries: Int
    let started_at: String?
    let completed_at: String?
    let processing_time_seconds: Int?
    let queue_id: String?
    let created_at: String
    let updated_at: String?
}

struct DesignJobListResponse: Codable {
    let jobs: [DesignJobResponse]
    let total: Int
    let page: Int
    let page_size: Int
}

// MARK: - Subscription Models
struct SubscriptionStatusResponse: Codable {
    let subscription: SubscriptionResponse
    let can_use_ai_job: Bool
    let quota_info: QuotaInfo
}

struct SubscriptionResponse: Codable {
    let id: Int
    let user_id: Int
    let plan: String
    let status: String
    let billing_provider: String?
    let provider_subscription_id: String?
    let current_period_start: String?
    let current_period_end: String?
    let canceled_at: String?
    let trial_end: String?
    let ai_job_quota: Int
    let ai_jobs_used: Int
    let ai_jobs_remaining: Int
    let metadata: String?
    let created_at: String
    let updated_at: String?
}

struct QuotaInfo: Codable {
    let quota: Int
    let used: Int
    let remaining: Int
    let percentage_used: Double
}

struct CancelSubscriptionRequest: Codable {
    let reason: String?
}

struct CancelSubscriptionResponse: Codable {
    let message: String
    let subscription: SubscriptionResponse
}

// MARK: - API Error
struct APIError: Codable, Error {
    let detail: String
}

File: android/app/src/main/java/com/desiora/ai/data/model/ApiModels.kt
package com.desiora.ai.data.model

import com.google.gson.annotations.SerializedName

// Authentication Models
data class LoginRequest(
    val email: String,
    val password: String
)

data class RegisterRequest(
    val email: String,
    val password: String,
    val username: String? = null,
    val full_name: String? = null
)

data class TokenResponse(
    @SerializedName("access_token") val accessToken: String,
    @SerializedName("refresh_token") val refreshToken: String,
    @SerializedName("token_type") val tokenType: String = "bearer"
)

data class RefreshTokenRequest(
    @SerializedName("refresh_token") val refreshToken: String
)

data class UserResponse(
    val id: Int,
    val email: String,
    val username: String?,
    @SerializedName("full_name") val fullName: String?,
    @SerializedName("is_active") val isActive: Boolean,
    @SerializedName("is_verified") val isVerified: Boolean,
    val role: String,
    @SerializedName("oauth_provider") val oauthProvider: String,
    @SerializedName("created_at") val createdAt: String,
    @SerializedName("updated_at") val updatedAt: String?
)

data class OAuthRequest(
    val code: String,
    val provider: String,
    @SerializedName("redirect_uri") val redirectUri: String? = null
)

// Image Models
data class PresignUploadRequest(
    val filename: String,
    @SerializedName("content_type") val contentType: String,
    @SerializedName("file_size") val fileSize: Long
)

data class PresignUploadResponse(
    @SerializedName("upload_url") val uploadUrl: String,
    @SerializedName("s3_key") val s3Key: String,
    @SerializedName("image_id") val imageId: Int,
    @SerializedName("expires_in") val expiresIn: Int
)

data class ConfirmUploadRequest(
    @SerializedName("s3_key") val s3Key: String
)

// Design Job Models
data class DesignJobCreateRequest(
    @SerializedName("job_type") val jobType: String,
    val prompt: String,
    val parameters: Map<String, Any>? = null
)

data class DesignJobResponse(
    val id: Int,
    @SerializedName("user_id") val userId: Int,
    @SerializedName("job_type") val jobType: String,
    val prompt: String,
    val status: String,
    val parameters: Map<String, Any>?,
    @SerializedName("result_urls") val resultUrls: List<String>?,
    @SerializedName("result_metadata") val resultMetadata: Map<String, Any>?,
    @SerializedName("error_message") val errorMessage: String?,
    @SerializedName("retry_count") val retryCount: Int,
    @SerializedName("max_retries") val maxRetries: Int,
    @SerializedName("started_at") val startedAt: String?,
    @SerializedName("completed_at") val completedAt: String?,
    @SerializedName("processing_time_seconds") val processingTimeSeconds: Int?,
    @SerializedName("queue_id") val queueId: String?,
    @SerializedName("created_at") val createdAt: String,
    @SerializedName("updated_at") val updatedAt: String?
)

data class DesignJobListResponse(
    val jobs: List<DesignJobResponse>,
    val total: Int,
    val page: Int,
    @SerializedName("page_size") val pageSize: Int
)

// Subscription Models
data class SubscriptionStatusResponse(
    val subscription: SubscriptionResponse,
    @SerializedName("can_use_ai_job") val canUseAiJob: Boolean,
    @SerializedName("quota_info") val quotaInfo: QuotaInfo
)

data class SubscriptionResponse(
    val id: Int,
    @SerializedName("user_id") val userId: Int,
    val plan: String,
    val status: String,
    @SerializedName("billing_provider") val billingProvider: String?,
    @SerializedName("provider_subscription_id") val providerSubscriptionId: String?,
    @SerializedName("current_period_start") val currentPeriodStart: String?,
    @SerializedName("current_period_end") val currentPeriodEnd: String?,
    @SerializedName("canceled_at") val canceledAt: String?,
    @SerializedName("trial_end") val trialEnd: String?,
    @SerializedName("ai_job_quota") val aiJobQuota: Int,
    @SerializedName("ai_jobs_used") val aiJobsUsed: Int,
    @SerializedName("ai_jobs_remaining") val aiJobsRemaining: Int,
    val metadata: String?,
    @SerializedName("created_at") val createdAt: String,
    @SerializedName("updated_at") val updatedAt: String?
)

data class QuotaInfo(
    val quota: Int,
    val used: Int,
    val remaining: Int,
    @SerializedName("percentage_used") val percentageUsed: Double
)

data class CancelSubscriptionRequest(
    val reason: String? = null
)

data class CancelSubscriptionResponse(
    val message: String,
    val subscription: SubscriptionResponse
)

// Admin Models
data class UserListResponse(
    val users: List<UserResponse>,
    val total: Int,
    val page: Int,
    @SerializedName("page_size") val pageSize: Int
)

data class AdminStatsResponse(
    val users: UserStats,
    val subscriptions: SubscriptionStats,
    val jobs: JobStats,
    val usage: UsageStats,
    @SerializedName("generated_at") val generatedAt: String
)

data class UserStats(
    val total: Int,
    val active: Int,
    val inactive: Int
)

data class SubscriptionStats(
    val total: Int,
    @SerializedName("by_plan") val byPlan: Map<String, Int>,
    @SerializedName("by_status") val byStatus: Map<String, Int>
)

data class JobStats(
    val total: Int,
    @SerializedName("by_status") val byStatus: Map<String, Int>,
    @SerializedName("average_per_user") val averagePerUser: Double
)

data class UsageStats(
    @SerializedName("ai_jobs_used") val aiJobsUsed: Int,
    @SerializedName("ai_jobs_quota") val aiJobsQuota: Int,
    @SerializedName("usage_percentage") val usagePercentage: Double,
    @SerializedName("top_users") val topUsers: List<TopUser>
)

data class TopUser(
    @SerializedName("user_id") val userId: Int,
    val email: String,
    val username: String?,
    @SerializedName("job_count") val jobCount: Int
)




